### PR TITLE
refactor: Migrate WarningsViewModel to React

### PR DESCRIPTION
**Pull request by thisisamir98**
telegram-cloud-photo-size-4-5882238633254172878-y

### CODE REVIEW

File: src/script/calling/CallingRepository.ts [hunk 1]
```diff
@@ -500,7 +500,7 @@ export class CallingRepository {
         this.leaveCall(activeCall.conversationId, LEAVE_CALL_REASON.USER_TURNED_UNVERIFIED);
         amplify.publish(
           WebAppEvents.WARNING.MODAL,
-          ModalsViewModel.TYPE.ACKNOWLEDGE,
+          WarningModalType.ACKNOWLEDGE,
```
Comment (przemvs) [inline src/script/calling/CallingRepository.ts]: I think amplify.publish should be changed for render modal function, and pass it to the DOM - like a Util/renderElement :)

File: src/script/components/Modals/WarningModal/WarningModal.tsx [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    if (typeof action === 'function') {
+      if (content?.currentType === WarningModalType.OPTION) {
+        return action(optionChecked);
+      }
+      if (content?.currentType === WarningModalType.INPUT) {
+        return action(inputValue);
+      }
+      if (content?.currentType === WarningModalType.PASSWORD) {
+        return action(passwordValue);
+      }
```
Comment (przemvs) [inline src/script/components/Modals/WarningModal/WarningModal.tsx]: I'll prepare object and check by content?.currentType if exist, eg.

```
const actions = {
  [WarningModalType.OPTION]: action(optionChecked),
  ...moreAcitons
}

content.currentType ? actions[content.currentType] : action()
```

WDYT ?

File: src/script/components/Modals/WarningModal/WarningModal.tsx [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    }
+  };
+
+  return (
+    <div id="modals">
+      <ModalComponent
+        isShown={isModalVisible}
+        onClosed={onModalHidden}
+        aria-describedby="modal-description"
+        onBgClick={() => onBgClick()}
```
Comment (przemvs) [inline src/script/components/Modals/WarningModal/WarningModal.tsx]:
```suggestion
        onBgClick={onBgClick}
```

File: src/script/components/Modals/WarningModal/WarningModal.tsx [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+                data-uie-name="do-close"
+              >
+                <Icon.Close className="modal__header__icon" aria-hidden="true" />
+              </button>
+            </div>
+            <div className="modal__body" ref={initFadingScrollbar}>
+              {(messageHtml || messageText) && (
+                <div className="modal__text" data-uie-name="status-modal-text">
+                  {messageHtml && <div id="modal-description" dangerouslySetInnerHTML={{__html: messageHtml}} />}
+                  {messageText && <div id="modal-description">{messageText}</div>}
```
Comment (przemvs) [inline src/script/components/Modals/WarningModal/WarningModal.tsx]: What if you will have messageHTML and messageText? Then you'll have duplicated id's. :)

File: src/script/components/Modals/WarningModal/WarningModal.tsx [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+                </div>
+              )}
+              {hasPassword && (
+                <form onSubmit={() => doAction(confirm, !!closeOnConfirm)}>
+                  <input
+                    className="modal__input"
+                    type="password"
+                    value={passwordValue}
+                    placeholder={inputPlaceholder ?? ''}
```
Comment (przemvs) [inline src/script/components/Modals/WarningModal/WarningModal.tsx]: You can define empty string in inputPlaceholder as default prop

File: src/script/components/Modals/WarningModal/WarningModal.tsx [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+                  />
+                </form>
+              )}
+              {hasInput && (
+                <form onSubmit={() => doAction(confirm, !!closeOnConfirm)}>
+                  <input
+                    maxLength={64}
+                    className="modal__input"
+                    value={inputValue}
+                    placeholder={inputPlaceholder ?? ''}
```
Comment (przemvs) [inline src/script/components/Modals/WarningModal/WarningModal.tsx]: You can define empty string in inputPlaceholder as default prop

File: src/script/components/Modals/WarningModal/WarningModal.tsx [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+                  />
+                </form>
+              )}
+              {errorMessage && <div className="modal__input__error">{errorMessage}</div>}
+              {hasOption && (
+                <div className="modal-option">
+                  <div className="checkbox accent-text">
```
Comment (przemvs) [inline src/script/components/Modals/WarningModal/WarningModal.tsx]: You can't join this 2 divs into 1 ? :)

File: src/script/components/Modals/WarningModal/WarningModal.tsx [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+                      data-uie-name="modal-option-checkbox"
+                    />
+                    <label className="label-xs" htmlFor="clear-data-checkbox">
+                      <span className="modal-option-text text-background">{checkboxLabel}</span>
+                    </label>
+                  </div>
+                </div>
+              )}
+              <div className={cx('modal__buttons', {'modal__buttons--column': hasMultipleSecondary})}>
+                {(Array.isArray(secondaryAction) ? secondaryAction : [secondaryAction]).map(action => (
```
Comment (przemvs) [inline src/script/components/Modals/WarningModal/WarningModal.tsx]: Please move this checking to variable, it will be more readable :)

File: src/script/components/Modals/WarningModal/WarningModal.tsx [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+                  <button
+                    key={`${action?.text}-${action?.uieName}`}
```
Comment (przemvs) [inline src/script/components/Modals/WarningModal/WarningModal.tsx]: If it doesn't containt action?.text and action?.uieName it will be 'undefined-undefined' :(

File: src/script/components/Modals/WarningModal/WarningModalState.ts [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+  titleText: '',
+};
+
+const logger = getLogger('WarningModalState');
+
+const useWarningModalState = create<WarningModalState>((set, get) => ({
+  addToQueue: (modalItem: ModalItem) =>
+    set(state => {
+      return {...state, queue: [...state.queue, modalItem]};
+    }),
```
Comment (przemvs) [inline src/script/components/Modals/WarningModal/WarningModalState.ts]: You can simplify it :)

File: src/script/components/Modals/WarningModal/WarningModalState.ts [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+  currentModalContent: defaultContent,
+  currentModalId: null,
+  errorMessage: null,
+  existsInQueue: (modalItem: ModalItem): boolean =>
+    get().queue.findIndex(queueItem => queueItem.id === modalItem.id) !== -1,
+  queue: [],
+  removeFirstItemInQueue: () =>
+    set(state => {
+      return {...state, queue: [...state.queue.slice(1)]};
+    }),
```
Comment (przemvs) [inline src/script/components/Modals/WarningModal/WarningModalState.ts]: You can simplify it :)

File: src/script/components/Modals/WarningModal/WarningModalState.ts [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+  replaceInQueue: (modalItem: ModalItem) =>
+    set(state => {
+      const updatedQueue = [...state.queue];
+      const index = updatedQueue.findIndex(queueItem => queueItem.id === modalItem.id);
+      updatedQueue[index] = modalItem;
```
Comment (przemvs) [inline src/script/components/Modals/WarningModal/WarningModalState.ts]: What do you think to use this (?) : 

```suggestion
      const updatedQueue = state.queue.map(queueItem => queueItem.id === modalItem.id ? modalItem : queueItem);
```

File: src/script/components/Modals/WarningModal/WarningModalState.ts [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+    set(state => ({...state, currentModalContent: nextCurrentModaContent})),
+  updateCurrentModalId: (nextCurrentModalId: string | null) =>
+    set(state => ({...state, currentModalId: nextCurrentModalId})),
+  updateErrorMessage: nextErrorMessage => set(state => ({...state, errorMessage: nextErrorMessage})),
+}));
+
+const onIncmoingModalEvent = (type: WarningModalType, options: ModalOptions, modalId = createRandomUuid()) => {
+  const {currentModalId, existsInQueue, addToQueue, replaceInQueue} = useWarningModalState.getState();
+
+  const alreadyOpen = modalId && modalId === currentModalId;
```
Comment (przemvs) [inline src/script/components/Modals/WarningModal/WarningModalState.ts]: It cannot be `const alreadyOpen = modalId === currentModalId;` ?

File: src/script/components/Modals/WarningModal/WarningModalState.ts [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+    onBgClick: preventClose ? noop : removeCurrentModal,
+    primaryAction: primaryAction ?? null,
+    secondaryAction: secondaryAction ?? null,
+    showClose,
+    titleText: text.title ?? null,
+  };
+
+  switch (type) {
+    case WarningModalType.ACCOUNT_NEW_DEVICES: {
+      content.titleText = t('modalAccountNewDevicesHeadline');
```
Comment (przemvs) [inline src/script/components/Modals/WarningModal/WarningModalState.ts]: I'll go with if && return instead of switch..case :)

File: src/script/components/Modals/WarningModal/WarningModalTypes.ts [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+  preventClose?: boolean;
+  /** Called when action in modal is triggered */
+  primaryAction?: Action;
+  /** Called when secondary action in modal is triggered */
+  secondaryAction?: Action[] | Action;
+  showClose?: boolean;
+  text?: Text;
+}
+
+export enum WarningModalType {
```
Comment (przemvs) [inline src/script/components/Modals/WarningModal/WarningModalTypes.ts]: And i think WarningModal it's bad name, because it's not only WarningModal ( if i good think ) :)

**Review by przemvs — CHANGES_REQUESTED** [review]
Try to remove all ampliy.publish/amplify.subscribe. Instead of this use renderElement maybe. I think we wanna remove all amplify functionality ?

File: src/script/calling/CallingRepository.ts (hunk 1 above)
Comment (atomrc) [reply src/script/calling/CallingRepository.ts]: 👍 +1 
Yes we should try to move out from `amplify` which is making a lot of interactions hard to follow and not typed

File: src/script/components/Modals/WarningModal/WarningModal.tsx (hunk 7 above)
Comment (thisisamir98) [reply src/script/components/Modals/WarningModal/WarningModal.tsx]: css styles break

File: src/script/components/Modals/WarningModal/WarningModalState.ts (hunk 14 above)
Comment (thisisamir98) [reply src/script/components/Modals/WarningModal/WarningModalState.ts]: function has more things to do so can not return and also it has cases handling same logic for different types which in if statement would be a log of || like if (type === WarningModalType.INPUT || type === ....)

File: src/script/components/Modals/WarningModal/WarningModal.tsx (hunk 2 above)
Comment (thisisamir98) [reply src/script/components/Modals/WarningModal/WarningModal.tsx]: https://github.com/wireapp/wire-webapp/pull/13748/commits/f209919f86a342b01b7bed87a75996e3b6e7069d

File: src/script/components/Modals/WarningModal/WarningModalTypes.ts (hunk 15 above)
Comment (thisisamir98) [reply src/script/components/Modals/WarningModal/WarningModalTypes.ts]: changed to PrimaryModal

File: src/script/calling/CallingRepository.ts (hunk 1 above)
Comment (thisisamir98) [reply src/script/calling/CallingRepository.ts]: https://github.com/wireapp/wire-webapp/pull/13748/commits/342f8a0afe9ebd679f413dcf01969fa0c5773b22

**Review by V-Gira — COMMENTED** [review]
Can't really contribute anything but I really like the move to zustand too, feels like I can understand the state management without massive headaches for a change :)

File: src/script/components/Modals/PrimaryModal/PrimaryModal.tsx [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+            </div>
+          </>
+        )}
+      </ModalComponent>
+    </div>
+  );
+};
+
+const PrimaryModal = {
+  add: addNewModalToQueue,
```
Comment (atomrc) [inline src/script/components/Modals/PrimaryModal/PrimaryModal.tsx]:
```suggestion
  show: addNewModalToQueue,
```

I think `show` makes a little more sense when we talk about modals. WDYT?

File: src/script/components/Modals/PrimaryModal/PrimaryModal.tsx [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ *
+ */
+
+import React, {useState} from 'react';
```
Comment (atomrc) [inline src/script/components/Modals/PrimaryModal/PrimaryModal.tsx:20]: Let's create a `index.ts` that exports the public API to the modal component. 
So that we can use it by doing `import ... from 'path/to/PrimaryModal` instead of `path/to/PrimaryModal/PrimaryModal`

File: src/script/components/Modals/PrimaryModal/PrimaryModal.tsx [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+import {Action, PrimaryModalType} from './PrimaryModalTypes';
+import {
+  usePrimaryModalState,
+  showNextModalInQueue,
+  defaultContent,
+  removeCurrentModal,
+  addNewModalToQueue,
+} from './PrimaryModalState';
+
+export interface PrimaryModalProps {}
```
Comment (atomrc) [inline src/script/components/Modals/PrimaryModal/PrimaryModal.tsx]: probably not needed :D

File: src/script/components/Modals/PrimaryModal/PrimaryModal.tsx [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+  );
+};
+
+const PrimaryModal = {
+  add: addNewModalToQueue,
+  init: () => {
+    renderElement<PrimaryModalProps>(PrimaryModalComponent)({});
+    showNextModalInQueue();
+  },
+  type: PrimaryModalType,
```
Comment (atomrc) [inline src/script/components/Modals/PrimaryModal/PrimaryModal.tsx]: we could export this in its own export and in the `index.ts` file

File: src/script/components/Modals/PrimaryModal/PrimaryModal.tsx [hunk 20]
```diff
@@ -0,0 +1,222 @@
...
+    inputPlaceholder,
+    messageHtml,
+    messageText,
+    modalUie,
+    onBgClick,
+    primaryAction,
+    secondaryAction,
+    titleText,
+    closeBtnTitle,
+  } = content || defaultContent;
```
Comment (atomrc) [inline src/script/components/Modals/PrimaryModal/PrimaryModal.tsx]: The default value should live only inside the state manager (the value it returns should be the default value and never be undefined if we don't expect it)

File: src/script/components/Modals/PrimaryModal/PrimaryModal.tsx [hunk 21]
```diff
@@ -0,0 +1,222 @@
...
+                  {messageText && <div id="modal-description-text">{messageText}</div>}
+                </div>
+              )}
+              {hasPassword && (
+                <form onSubmit={() => doAction(confirm, !!closeOnConfirm)}>
+                  <input
+                    className="modal__input"
+                    type="password"
+                    value={passwordValue}
+                    placeholder={inputPlaceholder ?? ''}
```
Comment (przemvs) [inline src/script/components/Modals/PrimaryModal/PrimaryModal.tsx]: inputPlaceholder doesn't have default value (empty string) ?

File: src/script/components/Modals/PrimaryModal/PrimaryModal.tsx [hunk 22]
```diff
@@ -0,0 +1,222 @@
...
+                  />
+                </form>
+              )}
+              {hasInput && (
+                <form onSubmit={() => doAction(confirm, !!closeOnConfirm)}>
+                  <input
+                    maxLength={64}
+                    className="modal__input"
+                    value={inputValue}
+                    placeholder={inputPlaceholder ?? ''}
```
Comment (przemvs) [inline src/script/components/Modals/PrimaryModal/PrimaryModal.tsx]: inputPlaceholder doesn't have default value (empty string) ?

File: src/script/components/Modals/PrimaryModal/PrimaryModal.tsx [hunk 23]
```diff
@@ -0,0 +1,222 @@
...
+                    />
+                    <label className="label-xs" htmlFor="clear-data-checkbox">
+                      <span className="modal-option-text text-background">{checkboxLabel}</span>
+                    </label>
+                  </div>
+                </div>
+              )}
+              <div className={cx('modal__buttons', {'modal__buttons--column': hasMultipleSecondary})}>
+                {secondaryActions
+                  .filter((action): action is Action => action !== null)
```
Comment (przemvs) [inline src/script/components/Modals/PrimaryModal/PrimaryModal.tsx]: Maybe move this to guards/Modal.ts (isAction) or smth like that :)

File: src/script/components/Modals/PrimaryModal/PrimaryModal.tsx [hunk 24]
```diff
@@ -0,0 +1,222 @@
...
+                  .map(action => (
+                    <button
+                      key={`${action?.text}-${action?.uieName}`}
```
Comment (przemvs) [inline src/script/components/Modals/PrimaryModal/PrimaryModal.tsx]: Key can be 'undefined-undefined' - i think it's wrong.

File: src/script/components/Modals/PrimaryModal/PrimaryModal.tsx [hunk 25]
```diff
@@ -0,0 +1,222 @@
...
+                      type="button"
+                      onClick={() => action?.action && doAction(action?.action, true, true)}
```
Comment (przemvs) [inline src/script/components/Modals/PrimaryModal/PrimaryModal.tsx]:
```suggestion
                      onClick={() => action?.action && doAction(action.action, true, true)}
```

File: src/script/components/Modals/PrimaryModal/PrimaryModal.tsx [hunk 26]
```diff
@@ -0,0 +1,222 @@
...
+                      data-uie-name={action?.uieName}
+                      className={cx('modal__button modal__button--secondary', {
+                        'modal__button--full': hasMultipleSecondary,
+                      })}
+                    >
+                      {action?.text}
```
Comment (przemvs) [inline src/script/components/Modals/PrimaryModal/PrimaryModal.tsx]: Maybe add custom text if action didn't have any text.

File: src/script/components/Modals/PrimaryModal/PrimaryModal.tsx [hunk 27]
```diff
@@ -0,0 +1,222 @@
...
+const PrimaryModal = {
+  add: addNewModalToQueue,
+  init: () => {
+    renderElement<PrimaryModalProps>(PrimaryModalComponent)({});
+    showNextModalInQueue();
+  },
+  type: PrimaryModalType,
+};
+
+export default PrimaryModal;
```
Comment (przemvs) [inline src/script/components/Modals/PrimaryModal/PrimaryModal.tsx]: I'm thinking about PrimaryModal name. for me it can be Modal - because it can be reused, PrimaryModal sounds like we have multiple Modal styles, or smth like that.

File: src/script/components/Modals/PrimaryModal/PrimaryModalState.ts [hunk 28]
```diff
@@ -0,0 +1,214 @@
...
+    showClose = false,
+    text = {} as Text,
+  } = options;
+  const content = {
+    checkboxLabel: text.option ?? null,
+    closeBtnTitle: text.closeBtnLabel,
+    closeFn: close,
+    closeOnConfirm,
+    currentType: type,
+    inputPlaceholder: text.input ?? null,
```
Comment (przemvs) [inline src/script/components/Modals/PrimaryModal/PrimaryModalState.ts]: I think it can be empty string instead of null :)

File: src/script/components/Modals/PrimaryModal/PrimaryModalState.ts [hunk 29]
```diff
@@ -0,0 +1,214 @@
...
+    data,
+    preventClose = false,
+    primaryAction,
+    secondaryAction,
+    hideSecondary,
+    showClose = false,
+    text = {} as Text,
+  } = options;
+  const content = {
+    checkboxLabel: text.option ?? null,
```
Comment (przemvs) [inline src/script/components/Modals/PrimaryModal/PrimaryModalState.ts]: I'll go with string types instead of null for string values :)

File: src/script/components/Modals/PrimaryModal/PrimaryModalTypes.ts [hunk 30]
```diff
@@ -0,0 +1,83 @@
...
+  currentType: string | null;
+  inputPlaceholder: string | null;
+  messageHtml: string | null;
+  messageText: string | null;
+  modalUie: string;
+  onBgClick: () => void;
+  primaryAction: Action | null;
+  secondaryAction: Action[] | Action | null;
+  titleText: string | null;
+  closeBtnTitle?: string;
```
Comment (przemvs) [inline src/script/components/Modals/PrimaryModal/PrimaryModalTypes.ts]: I'll go for string types as optional and without null. We can pass empty string for string :)

**Review by przemvs — CHANGES_REQUESTED** [review]
Good job with replacing amplify!

Few comments more :)

File: src/script/components/Modals/PrimaryModal/PrimaryModal.tsx (hunk 23 above)
Comment (thisisamir98) [reply src/script/components/Modals/PrimaryModal/PrimaryModal.tsx]: this is a bit too much just for a null check

File: src/script/components/Modals/PrimaryModal/PrimaryModal.test.tsx [hunk 31]
```diff
@@ -0,0 +1,96 @@
...
+
+class PrimaryModalPage extends TestPage<{}> {
+  constructor() {
+    super(PrimaryModalComponent);
+  }
+
+  getWrapperElement = () => this.get('div#modals');
+  getPrimaryActionButton = () => this.get('[data-uie-name="do-action"]');
+  getSecondaryActionButton = () => this.get('[data-uie-name="do-secondary"]');
+}
```
Comment (atomrc) [inline src/script/components/Modals/PrimaryModal/PrimaryModal.test.tsx]: Let's use `@testing-library` directly and all it's super nice primitives (like `getByTestId`). 
You can inspire from what @PatrykBuniX has been doing lately migrating a few tests to testling-lib (cf https://github.com/wireapp/wire-webapp/pull/13751/files#diff-aa9cafe9a5b4c4728e53a5935e604adaf6c345a38a55778469861301792662e0R46)

**Review by przemvs — APPROVED** [review]
Code LGTM :)

**Review by atomrc — APPROVED** [review]

**Review by PatrykBuniX — APPROVED** [review]
💪
